### DISCUSSION TITLE
Why am I getting hydration error while using getLayout

### DISCUSSION THREAD

**Discussion by MrtYlnzz** [vercel/next.js#50594] [Help] [answered]
### Summary

Basically I am using per page layout pattern mentioned [here](https://nextjs.org/docs/pages/building-your-application/routing/pages-and-layouts#per-page-layouts) and it causes [hydration error](https://github.com/vercel/next.js/assets/6218537/bdf17618-2172-420b-a671-f4349885360b) whenever I use one component in different routes. 

For example I have an about page and I import it inside 2 other directories inside pages folder and export it again with different layouts by using getLayout on each route. This causes hydration error whenever navigating between these pages couple of times in both development and production.

https://github.com/vercel/next.js/assets/6218537/d579c71d-6fe0-4b5c-8402-97e4110d4874

Exporting it as a new component solves this issue completely as below;

```typescript
const BlueAbout = () => {
  return (
    <BlueLayout>
      <About />
    </BlueLayout>
  );
};

export default BlueAbout;
``` 

What am I doing wrong?

### Example

https://github.com/MrtYlnzz/nextjs-layout

**Comment by icyJoseph** [comment, marked as answer]
I think the problem here is mutation.

The About page is,  a single instance exposed by your module.

Then at every page level, you mutate its getLayout function, which should lead to unpredictable behaviour.

That being said, I could not reproduce this reliably, like build and start the page, no navigation between pages. I saw it only a couple of times. What did trigger it every time was setting the CPUs to 1. So my guess is that if the page building is split between many workers, we won't really see the issue, I think? Not sure if SPA navigations might trigger it 🙈

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  experimental: {
    cpus: 1,
  },
};

module.exports = nextConfig;
```

Then the pages are all over the place, blue is red with no warning, red and static throw hydration errors...

I think, it is best to stay away from mutations, and do as you already figured out, create a page level component, and attach the getLayout function to it instead.

**Reply by MrtYlnzz** [reply]
Mutation makes sense, that is what I had assumed too, I am just not sure how that happens while About component is imported each time, shouldn't that create new About instance? Also I am not sure how workers are involved during production build since for example I've seen red/about somehow having BlueLayout inside its static html, this doesnt always happen like you've mentioned already.

**Reply by icyJoseph** [reply]
Importing doesn't create a new instance. 

For example, take this file: https://github.com/vercel/swr/blob/main/_internal/utils/timestamp.ts it creates an ever increasing counter, every time it is called in the various places where it is included, it mutates its internal counter.

In general people use this for various things, like global stores, look up tables, for mutation, etc, etc. 

You can think of its, as if, pages A, B, C all import module Z, so when bundled, the app places Z into a common to all pages group. Now everyone sees the same. It is not exactly the reason, but it is a good mental model.

You can try this out yourself, `yarn init` a project, install TypeScript for example, and create three files:

```ts
// utils.ts
// Global timestamp.
let __timestamp = 0

export const getTimestamp = () => ++__timestamp
```

And

```ts
// bar.ts
import { getTimestamp } from "./utils";

export const counter = getTimestamp;

```

Finally:

```ts
// foo.ts
import { getTimestamp } from "./utils";
import { counter } from "./bar";

console.log(counter === getTimestamp);
```

When you run and build this project, you get `true`.

**Reply by MrtYlnzz** [reply]
Totally forgot about that, I am not sure why I thought components would be any different than that, dumb me, thank you very much for detailed explanation. I think I will go with new app router since it is stable now and it comes with new layout structure that satisfies my use case.

**Reply by icyJoseph** [reply]
No worries. Please don't forget to mark as answered, and close the discussion.
